### PR TITLE
Allow parsing same-type bindables

### PR DESCRIPTION
**Pull request by frenzibyte**
Fixes a case osu!-side in testing where invoking `ToMod()` on a non-deserialized (manually constructed) `APIMod` instance would attempt to parse a `Bindable`, and would result in an "Type must be IConvertible" exception because of falling to the `default` case in the parse method. Upon discussing, parsing `Bindable`s sounds OK to do.

### CODE REVIEW

**Comment by bdach** [comment]
Needs test coverage.

**Comment by frenzibyte** [comment]
Added, that actually made me realize there's no direct check for different-type bindables. 😄

File: osu.Framework/Bindables/Bindable.cs [hunk 1]
```diff
@@ -245,6 +245,13 @@ public virtual void Parse(object input)
                     Value = t;
                     break;
 
+                case IBindable _:
+                    if (!(input is IBindable<T> bindable))
+                        throw new ArgumentException($"Attempting to parse {input.GetType()}, which has differnet value type from {typeof(T)}", nameof(input));
```
Comment (bdach) [inline osu.Framework/Bindables/Bindable.cs:250]: has typo, but in general i'd recommend writing this as

```suggestion
                        throw new ArgumentException($"Expected bindable of type {nameof(IBindable)}<{typeof(T)}>, got {input.GetType()}", nameof(input));
```

File: osu.Framework.Tests/Bindables/BindableTest.cs [hunk 2]
```diff
@@ -56,6 +56,25 @@ public void TestBindViaBindTarget()
...
+        [Test]
+        public void TestParseSameTypeBindable()
+        {
+            var bindable1 = new BindableInt();
+            var bindable2 = new BindableDouble();
+            var bindable3 = new BindableBool();
+
+            bindable1.Parse(new BindableInt(3));
+            bindable2.Parse(new BindableDouble(2.5));
+            bindable3.Parse(new BindableBool(true));
```
Comment (bdach) [inline osu.Framework.Tests/Bindables/BindableTest.cs:68]: should also cover not-exactly-same-but-convertible types (`Bindable<int>` and `BindableInt`, etc.), and also things like `Bindable<string>`

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Framework.Tests/Bindables/BindableTest.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Framework.Tests/Bindables/BindableTest.cs:69]: Does this fulfill what you're suggesting? https://github.com/ppy/osu-framework/pull/4110/commits/1314d13a749de26db437288dec60fe6f96ada6f5

**Review by bdach — APPROVED** [review]
looks fine to me

**Review by peppy — APPROVED** [review]
logical

**Pull request merged**